In [1]:
import pandas as pd
import numpy as np

In [5]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")
category = pd.read_csv("../data/raw/product_category_name_translation.csv")

In [6]:
print("Datasets loaded successfully!")

Datasets loaded successfully!


In [7]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [8]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in date_columns:
    orders[col] = pd.to_datetime(orders[col])

In [9]:
orders.dtypes

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [10]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [11]:
orders[
    orders["order_delivered_customer_date"].isnull()
]["order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

## Handling Missing Values - Orders

The missing values in delivery-related columns were analyzed before cleaning.

Most missing timestamps belong to orders with statuses such as **canceled**, **processing**, **shipped**, and **unavailable**, where missing delivery information is expected as part of the business process.

Since these missing values represent real operational scenarios rather than data quality issues, no imputation or deletion was performed.

In [12]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [16]:
products["product_weight_g"] = products["product_weight_g"].fillna(products["product_weight_g"].median())

products["product_length_cm"] = products["product_length_cm"].fillna(products["product_length_cm"].median())

products["product_height_cm"] = products["product_height_cm"].fillna(products["product_height_cm"].median())

products["product_width_cm"] = products["product_width_cm"].fillna(products["product_width_cm"].median())

# Verify
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                0
product_length_cm               0
product_height_cm               0
product_width_cm                0
dtype: int64

In [17]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                0
product_length_cm               0
product_height_cm               0
product_width_cm                0
dtype: int64

In [20]:
datasets = {
    "Customers": customers,
    "Orders": orders,
    "Order Items": order_items,
    "Payments": payments,
    "Reviews": reviews,
    "Products": products,
    "Sellers": sellers,
    "Geolocation": geolocation,
    "Category Translation": category
}

for name, df in datasets.items():
    print(f"{name}: {df.duplicated().sum()} duplicate rows")

Customers: 0 duplicate rows
Orders: 0 duplicate rows
Order Items: 0 duplicate rows
Payments: 0 duplicate rows
Reviews: 0 duplicate rows
Products: 0 duplicate rows
Sellers: 0 duplicate rows
Geolocation: 261831 duplicate rows
Category Translation: 0 duplicate rows


## Duplicate Records

A duplicate check was performed on all datasets.

- Customers, Orders, Order Items, Payments, Reviews, Products, Sellers, and Category Translation contained **no duplicate records**.
- The Geolocation dataset contains **261,831 duplicate rows**.

The geolocation duplicates were not removed during this stage because repeated geographic records can be valid and may be useful for future location-based analysis. Since this project focuses on sales analytics, the dataset is preserved in its original form.

In [21]:
customers.to_csv("../data/cleaned/customers_cleaned.csv", index=False)
orders.to_csv("../data/cleaned/orders_cleaned.csv", index=False)
order_items.to_csv("../data/cleaned/order_items_cleaned.csv", index=False)
payments.to_csv("../data/cleaned/payments_cleaned.csv", index=False)
reviews.to_csv("../data/cleaned/reviews_cleaned.csv", index=False)
products.to_csv("../data/cleaned/products_cleaned.csv", index=False)
sellers.to_csv("../data/cleaned/sellers_cleaned.csv", index=False)
geolocation.to_csv("../data/cleaned/geolocation_cleaned.csv", index=False)
category.to_csv("../data/cleaned/category_translation_cleaned.csv", index=False)

print("✅ All cleaned datasets have been saved successfully!")

✅ All cleaned datasets have been saved successfully!
